[crowdstrike] Paginate indicator requests

gsilvapt opened the pull request:
### Proposed changes

* `IndicatorsAPI.get_combined_indicator_entities` now accepts a `next_page` continuation token and re-exposes the next-page token of the current response as `response_body["next_page"]` (a single string, not a parsed query-string dict, so the contract is unambiguous).
* `IndicatorImporter._fetch_indicators` is rewritten around a `_paginated_query_indicators` helper that walks every page returned by the API while honouring a configurable `max_records_per_run` cap. The cap is enforced at the resource level so a single page is never partially yielded with a dangling truncation, and pagination is robust to `meta.pagination` being missing or `None`.
* New `CROWDSTRIKE_INDICATOR_MAX_RECORDS_PER_RUN` setting (default `20000`) bounds the per-run volume. CrowdStrike can return up to ~1.5M IoCs over 30 days which would overwhelm the OpenCTI ingestion pipeline; set to `0` (or a negative number) to disable the cap entirely.
* The setting is exposed in `config.yml.sample`, `__metadata__/connector_config_schema.json` and `__metadata__/CONNECTOR_CONFIG_DOC.md`.
* New `tests/test_indicator_pagination.py` (11 unit tests) covers multi-page walking, stop-on-empty, the `meta.pagination == None` edge case, the per-run cap (within a single page and across multiple pages), the disabled cap (`None` / `0` / `-1`) and the `get_next_page` token parser.

### Related issues

* Closes #4497

### Checklist

- [x] I consider the submitted work as finished
- [x] I have signed my commits using GPG key.
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

PR rebased onto the current master (the connector layout evolved while this PR was open: `connector.py`, `RelatedActorImporter`, `AttackTechniqueLookup`, `indicator_max_age_by_type`, …) and squashed into a single GPG-signed commit. Every Copilot thread is addressed and resolved, and the per-run cap requested by @Powlinett is now wired through the pydantic settings, the env-var schema and the YAML config sample.

Local validation:
- `isort==6.0.1 --profile black` (clean)
- `black==25.1.0 --check` (clean)
- `flake8 --ignore=E,W` (clean)
- `pylint --disable=all --enable=no_generated_id_stix,no-value-for-parameter,unused-import` with the `linter_stix_id_generator` plugin → 10/10
- `pytest external-import/crowdstrike/tests/` → **141 passed**

gsilvapt commented:
> Hi @gsilvapt ,
> 
> Thank you for your contribution 🙏
> 
> I tested the pagination and it seems to work as expected 👍 Could you resolve the conflict on the PR so I can review it one last time and merge it please? 🙏
> 
> Feel free to ping me for any help 😇

Thank you for reviewing! Apologies for the late response.
Just solved the merge conflicts for a final round of tests 🙌 

Let me know if you need any further help.

SamuelHassine committed:
[crowdstrike] Paginate indicator requests and cap them per run

CrowdStrike's ``query_indicator_entities`` endpoint exposes a ``Next-Page``
HTTP header for deep pagination. Without using it, the connector was
limited to the API's per-call hard cap (10000 entities) and silently
dropped everything beyond that on each run. This commit walks every page
returned by the API up to a new configurable per-run cap so the ingestion
pipeline can keep up.

Highlights:
- ``IndicatorsAPI.get_combined_indicator_entities`` now accepts a
  ``next_page`` continuation token and re-exposes the next-page token of
  the current response as ``response_body["next_page"]``, a single string
  (not a parsed query-string dict) so the contract is unambiguous.
- ``IndicatorsAPI.get_next_page`` returns ``Optional[str]`` instead of a
  ``parse_qs`` dict; missing / empty tokens are normalised to ``None``.
- ``IndicatorImporter._fetch_indicators`` is rewritten around a
  ``_paginated_query_indicators`` helper that walks every page until the
  upstream API returns no more, while honouring a configurable
  ``max_records_per_run`` cap. The cap is enforced at the resource level
  so a single page is never partially yielded with a dangling truncation,
  and pagination is robust to ``meta.pagination`` being missing or
  ``None``.
- New ``CROWDSTRIKE_INDICATOR_MAX_RECORDS_PER_RUN`` setting (default
  ``20000``) bounds the per-run volume. CrowdStrike can return up to
  ~1.5M IoCs over 30 days which would overwhelm the OpenCTI ingestion
  pipeline; set to ``0`` (or a negative number) to disable the cap.
- The setting is exposed in ``config.yml.sample``,
  ``__metadata__/connector_config_schema.json`` and
  ``__metadata__/CONNECTOR_CONFIG_DOC.md``.
- New ``tests/test_indicator_pagination.py`` (11 unit tests) covers
  multi-page walking, stop-on-empty, the ``meta.pagination == None``
  edge case, the per-run cap (within a single page and across multiple
  pages), the disabled cap (``None`` / ``0`` / ``-1``) and the
  ``get_next_page`` token parser.

Validated locally:
- ``isort==6.0.1 --profile black`` (clean)
- ``black==25.1.0 --check`` (clean)
- ``flake8 --ignore=E,W`` (clean)
- ``pylint --enable=no_generated_id_stix,no-value-for-parameter,unused-import``
  with the ``linter_stix_id_generator`` plugin -> 10/10
- ``pytest external-import/crowdstrike/tests/`` -> 141 passed

Fixes #4497

SamuelHassine commented on external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py:
```diff
@@ -90,82 +90,136 @@ def run(self, state: Dict[str, Any]) -> Dict[str, Any]:
+                next_page_params["next_page"][0]
+                if isinstance(next_page_params["next_page"], list)
+                else next_page_params["next_page"]
             )
```
Fixed: the client now returns `next_page` as a single string token (not a `parse_qs` dict), so there is no ambiguity left. The importer forwards that token as-is on the next call and stops when it is `None` / empty.

SamuelHassine commented on external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py:
```diff
@@ -90,82 +90,136 @@ def run(self, state: Dict[str, Any]) -> Dict[str, Any]:
+                break
+
+            next_page_params = next_page_details
+
```
Done: `tests/test_indicator_pagination.py` (11 unit tests) covers multi-page walking, stopping when the next-page token is missing, the per-run cap (within a single page and across multiple pages), the disabled-cap variants, the `meta.pagination is None` edge case, and the `get_next_page` token parser. All 141 tests in `external-import/crowdstrike/tests/` pass locally.

SamuelHassine commented on external-import/crowdstrike/src/crowdstrike_feeds_services/client/indicators.py:
```diff
@@ -9,19 +10,33 @@ def __init__(self, helper):
+        response = self.cs_intel.query_indicator_entities(**kwargs)
 
         response_body = response["body"]
         response_body["next_page_details"] = None
```
Fixed: `get_combined_indicator_entities` now sets `response_body['next_page']` to a single string token (the value of the `next_page` query parameter extracted from the upstream `Next-Page` header) or `None`. `IndicatorsAPI.get_next_page` is typed `Optional[str]` accordingly. Callers just need to forward that string back as the `next_page=` kwarg, which is what `_paginated_query_indicators` does.

SamuelHassine commented on external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py:
```diff
@@ -90,82 +90,136 @@ def run(self, state: Dict[str, Any]) -> Dict[str, Any]:
+
+            # Log progress
+            meta = response.get("meta", {})
+            pagination = meta.get("pagination", {})
```
Fixed: `_paginated_query_indicators` now uses `meta.get('pagination') or {}` (and `meta = response.get('meta') or {}` for good measure), so the loop is safe whether `meta` / `pagination` is missing, `None` or an empty dict. There is also a dedicated unit test (`test_pagination_handles_missing_pagination_metadata`) covering the `meta.pagination == None` case.

SamuelHassine commented:
@Powlinett The max-IoCs-per-run threshold is now in. `CROWDSTRIKE_INDICATOR_MAX_RECORDS_PER_RUN` defaults to `20000` (your suggested value) but can be tuned via env / `config.yml` and disabled with `0` for users who want every IoC. Pagination logic refactored accordingly, all four Copilot threads addressed, and the change is covered by 11 new unit tests in `tests/test_indicator_pagination.py` (full suite still passes: 141 tests).

SamuelHassine reviewed (dismissed):
Full review done.

Rebased the change set onto the current master (the connector layout has evolved significantly since this PR was opened: new `connector.py`, `RelatedActorImporter`, `AttackTechniqueLookup`, `indicator_max_age_by_type`, …) and squashed everything into a single GPG-signed commit. All four Copilot review threads have been addressed and resolved, the per-run cap requested by @Powlinett is now wired through, and the change is now covered by 11 new unit tests.

Highlights:
- `IndicatorsAPI.get_combined_indicator_entities` now accepts a `next_page` continuation token and re-exposes the next-page token of the current response as `response_body["next_page"]` (a single string, not a parsed query-string dict). `get_next_page` is typed `Optional[str]` accordingly.
- `IndicatorImporter._fetch_indicators` is rewritten around `_paginated_query_indicators`, which walks every page until either the upstream API returns no more or the configured cap is reached. The cap is enforced at the resource level (never partial-page truncation), and the loop is robust to `meta.pagination` being missing or `None`.
- New `CROWDSTRIKE_INDICATOR_MAX_RECORDS_PER_RUN` setting (default `20000`) bounds the per-run volume; set to `0` (or a negative number) to disable the cap entirely. Exposed in `config.yml.sample`, the JSON schema and the auto-generated docs.
- New `tests/test_indicator_pagination.py` (11 unit tests) covers the multi-page walk, stop-on-empty, the `meta.pagination == None` edge case, the per-run cap (within a single page and across multiple pages), the three disabled-cap variants (`None` / `0` / `-1`) and the `get_next_page` token parser.

CI is fully green on the GitHub Actions side (`Test external-import/crowdstrike`, `Test tests/test-requirements.txt`, all `PR conventions checks`, `Baseline coverage`, `Do Not Merge label check`, `codecov/patch`, `codecov/project`). The CircleCI workflows haven't fired yet on this revision but all 141 tests in `external-import/crowdstrike/tests/` pass locally with `isort==6.0.1` / `black==25.1.0` / `flake8` / `pylint --enable=no_generated_id_stix,no-value-for-parameter,unused-import` clean (pylint 10/10).

The only remaining unstable status is `filigran/cla` (waiting for @gsilvapt to sign), which can be overridden at merge time.

gsilvapt commented:
Just signed it to unblock you @SamuelHassine , even though you can overwrite it apparently. Thanks for bumping this!

SamuelHassine committed:
[crowdstrike] Address remaining review comments

* `indicator/importer.py`: drop the unused ``self.next_page`` attribute
  (leftover from the previous pagination implementation - the actual
  cursor lives in the local ``next_page_token`` inside
  ``_paginated_query_indicators``).
* `indicator/importer.py`: update the ``_paginated_query_indicators``
  docstring to reflect the real behavior - the cap is enforced at the
  resource level by slicing the *last* page to exactly the remaining
  quota, so the aggregated batch contains at most
  ``max_records_per_run`` records and we never yield a record beyond
  the cap.
* `config.yml.sample`: align the ``indicator_max_records_per_run`` inline
  comment with `__metadata__/CONNECTOR_CONFIG_DOC.md` and the JSON
  schema by mentioning that a negative value also disables the cap
  (``Set to 0 (or any negative number) to disable the cap``).

SamuelHassine commented on external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py:
```diff
@@ -91,6 +92,11 @@ def __init__(self, config: IndicatorImporterConfig) -> None:
         self.scopes = config.scopes
```
Fixed: removed the unused `self.next_page` attribute from `IndicatorImporter.__init__`. It was leftover from the previous pagination implementation - the actual cursor lives in the local `next_page_token` inside `_paginated_query_indicators`, and nothing else read this attribute. `pylint --enable=unused-import` and the existing 141-test suite still pass.

SamuelHassine commented on external-import/crowdstrike/src/crowdstrike_feeds_connector/indicator/importer.py:
```diff
@@ -138,54 +144,117 @@ def run(self, state: Dict[str, Any]) -> Dict[str, Any]:
+        ``max_records_per_run`` is set, pagination stops as soon as that many
+        indicators have been collected; the cap is enforced at the resource
+        level so a single page is never partially yielded with a dangling
+        truncation.
```
Fixed: the docstring of `_paginated_query_indicators` has been rewritten to match the implementation. It now states that the cap is enforced at the resource level by slicing the *last* page to exactly the remaining quota, so the aggregated batch contains at most `max_records_per_run` records and we never yield a record beyond the cap. Slicing only happens on the page that crosses the cap boundary; previous pages are always returned in full, which is what `test_pagination_truncates_within_a_single_page` / `test_pagination_truncates_across_multiple_pages` already verify.

SamuelHassine commented on external-import/crowdstrike/src/config.yml.sample:
```diff
@@ -45,5 +45,6 @@ crowdstrike:
   indicator_high_score: 80
   indicator_high_score_labels: 'MaliciousConfidence/High'
   indicator_unwanted_labels: '' # Can be used to filter low confidence indicators: "MaliciousConfidence/Low", "MaliciousConfidence/Medium"
+  indicator_max_records_per_run: 20000 # Max indicators ingested per run (CrowdStrike can return up to ~1.5M IoCs over 30 days). Set to 0 to disable the cap.
```
Fixed: the inline comment in `src/config.yml.sample` now reads `Set to 0 (or any negative number) to disable the cap.`, matching what `__metadata__/CONNECTOR_CONFIG_DOC.md` and `__metadata__/connector_config_schema.json` already say. So all three docs are now consistent with the runtime behavior (`config.max_records_per_run if config.max_records_per_run and config.max_records_per_run > 0 else None` in the importer).

SamuelHassine reviewed (dismissed):
Re-approving after pushing review-comment cleanup (removed unused self.next_page, aligned docstring with slicing behavior, mentioned negative-cap-disables in config.yml.sample). All 7 review threads are resolved; locally validated with isort/black/flake8, pylint+linter_stix_id_generator (10/10), and pytest external-import/crowdstrike/tests/ (141 passed).

SamuelHassine committed:
[crowdstrike] Replace os.environ.setdefault with an autouse env fixture in pagination tests

The previous module-level ``os.environ.setdefault(...)`` calls in
``tests/test_indicator_pagination.py`` had two problems flagged by
review:

* ``setdefault`` does NOT override potentially-invalid values inherited
  from the test runner environment;
* the changes were never reverted, so they leaked into the other
  CrowdStrike test modules (most of which use ``mock_env_vars(...)``
  from ``conftest.py`` to patch + restore env).

The fix:

* a function-scoped ``autouse`` fixture (``_crowdstrike_env``) drives
  the env through ``monkeypatch.setenv``, which always overrides
  (not "set-if-unset") and is automatically restored by pytest at
  teardown — env can no longer leak across tests;
* the ``crowdstrike_feeds_connector`` / ``crowdstrike_feeds_services``
  imports are now lazy (inside ``_build_importer`` and the three
  ``test_get_next_page_*`` cases) so the env fixture has applied
  before the connector's pydantic settings are constructed at module
  import time.

All 141 tests in ``external-import/crowdstrike/tests/`` continue to
pass locally.

SamuelHassine commented on external-import/crowdstrike/tests/test_indicator_pagination.py:
```diff
@@ -0,0 +1,218 @@
+os.environ.setdefault("CONNECTOR_DURATION_PERIOD", "PT30M")
+os.environ.setdefault("CROWDSTRIKE_CLIENT_ID", "ChangeMe")
+os.environ.setdefault("CROWDSTRIKE_CLIENT_SECRET", "ChangeMe")
+
```
Fixed: the module-level `os.environ.setdefault(...)` block is gone. The env vars are now driven by a function-scoped `autouse` fixture (`_crowdstrike_env`) that uses `monkeypatch.setenv` ? which always **overrides** any value the test runner might have inherited (whereas `setdefault` would have silently kept an invalid one) ? and pytest restores the original env at teardown so the settings can no longer leak into the other CrowdStrike test modules.

Concretely:

```python
_REQUIRED_ENV: dict[str, str] = {
    "OPENCTI_URL": "http://localhost:8080",
    ...
}

@pytest.fixture(autouse=True)
def _crowdstrike_env(monkeypatch: pytest.MonkeyPatch) -> None:
    for key, value in _REQUIRED_ENV.items():
        monkeypatch.setenv(key, value)
```

The `crowdstrike_feeds_connector` / `crowdstrike_feeds_services` imports are also now **lazy** (inside `_build_importer` and the three `test_get_next_page_*` cases) so the env fixture has applied before the connector's pydantic settings are constructed at module import time.

All 141 tests in `external-import/crowdstrike/tests/` continue to pass locally.

SamuelHassine reviewed (approved):
Re-approving after replacing the module-level os.environ.setdefault block with an autouse monkeypatch.setenv fixture (overrides invalid runner values + cleans up at teardown) and lazy-importing the connector modules inside _build_importer so the fixture has applied before pydantic-settings runs. All 141 tests in external-import/crowdstrike/tests/ still pass; isort/black/flake8 clean.
